File: vigia_ui/lib/data/services/whep_live_session.dart
import 'dart:async';
import 'dart:math' as math;

import 'package:dio/dio.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter_webrtc/flutter_webrtc.dart';

enum WhepLiveStatus { connecting, playing, error }

/// Manages a receive-only WHEP session against MediaMTX.
///
/// Dispose closes the peer connection only — callers must not send
/// STOP_STREAMING here (MediaMTX runOnUnDemand owns that).
class WhepLiveSession extends ChangeNotifier {
  WhepLiveSession({required this.whepUrl, Dio? dio})
    : _dio = dio ?? Dio();

  final String whepUrl;
  final Dio _dio;

  final RTCVideoRenderer renderer = RTCVideoRenderer();

  WhepLiveStatus status = WhepLiveStatus.connecting;
  String? errorMessage;
  bool isPaused = false;

  RTCPeerConnection? _peerConnection;
  MediaStream? _remoteStream;
  bool _disposed = false;
  bool _rendererReady = false;
  int _connectGeneration = 0;

  Future<void> initialize() async {
    if (_rendererReady) return;
    await renderer.initialize();
    _rendererReady = true;
  }

  void beginConnecting() {
    if (_disposed) return;
    status = WhepLiveStatus.connecting;
    errorMessage = null;
    isPaused = false;
    notifyListeners();
  }

  void markError(Object error) {
    if (_disposed) return;
    status = WhepLiveStatus.error;
    errorMessage = error.toString();
    notifyListeners();
  }

  Future<void> connect({
    int maxAttempts = 24,
    Duration initialDelay = const Duration(milliseconds: 500),
    Duration maxDelay = const Duration(seconds: 5),
  }) async {
    if (_disposed) return;

    final generation = ++_connectGeneration;
    beginConnecting();

    var delay = initialDelay;
    Object? lastError;

    for (var attempt = 0; attempt < maxAttempts; attempt++) {
      if (_disposed || generation != _connectGeneration) return;

      try {
        await _negotiate();
        if (_disposed || generation != _connectGeneration) return;
        status = WhepLiveStatus.playing;
        errorMessage = null;
        notifyListeners();
        return;
      } catch (e) {
        lastError = e;
        await _teardownPeer();
        if (_disposed || generation != _connectGeneration) return;
        if (attempt == maxAttempts - 1) break;
        await Future<void>.delayed(delay);
        final nextMs = math.min(
          delay.inMilliseconds * 2,
          maxDelay.inMilliseconds,
        );
        delay = Duration(milliseconds: nextMs);
      }
    }

    if (_disposed || generation != _connectGeneration) return;
    markError(lastError ?? 'Failed to connect to live stream');
  }

  void togglePause() {
    final stream = _remoteStream;
    if (stream == null || status != WhepLiveStatus.playing) return;

    isPaused = !isPaused;
    for (final track in stream.getVideoTracks()) {
      track.enabled = !isPaused;
    }
    for (final track in stream.getAudioTracks()) {
      track.enabled = !isPaused;
    }
    notifyListeners();
  }

  Future<void> _negotiate() async {
    await initialize();

    final pc = await createPeerConnection({
      'sdpSemantics': 'unified-plan',
      'iceServers': [
        {'urls': 'stun:stun.l.google.com:19302'},
      ],
    });
    _peerConnection = pc;

    pc.onTrack = (RTCTrackEvent event) {
      if (_disposed) return;
      if (event.streams.isEmpty) return;
      _remoteStream = event.streams.first;
      renderer.srcObject = _remoteStream;
      notifyListeners();
    };

    await pc.addTransceiver(
      kind: RTCRtpMediaType.RTCRtpMediaTypeVideo,
      init: RTCRtpTransceiverInit(direction: TransceiverDirection.RecvOnly),
    );
    await pc.addTransceiver(
      kind: RTCRtpMediaType.RTCRtpMediaTypeAudio,
      init: RTCRtpTransceiverInit(direction: TransceiverDirection.RecvOnly),
    );

    final offer = await pc.createOffer();
    await pc.setLocalDescription(offer);
    await _waitForIceGathering(pc);

    final local = await pc.getLocalDescription();
    final sdp = local?.sdp;
    if (sdp == null || sdp.isEmpty) {
      throw Exception('Empty local SDP offer');
    }

    final response = await _dio.post<String>(
      whepUrl,
      data: sdp,
      options: Options(
        contentType: 'application/sdp',
        responseType: ResponseType.plain,
        headers: {'Accept': 'application/sdp'},
        validateStatus: (status) => status != null && status < 500,
      ),
    );

    final code = response.statusCode ?? 0;
    if (code != 200 && code != 201) {
      throw Exception('WHEP negotiate failed ($code)');
    }

    final answerSdp = response.data;
    if (answerSdp == null || answerSdp.isEmpty) {
      throw Exception('Empty WHEP SDP answer');
    }

    await pc.setRemoteDescription(
      RTCSessionDescription(answerSdp, 'answer'),
    );
  }

  Future<void> _waitForIceGathering(RTCPeerConnection pc) async {
    if (pc.iceGatheringState ==
        RTCIceGatheringState.RTCIceGatheringStateComplete) {
      return;
    }

    final completer = Completer<void>();
    pc.onIceGatheringState = (RTCIceGatheringState state) {
      if (state == RTCIceGatheringState.RTCIceGatheringStateComplete &&
          !completer.isCompleted) {
        completer.complete();
      }
    };

    await completer.future.timeout(
      const Duration(seconds: 8),
      onTimeout: () {},
    );
  }

  Future<void> _teardownPeer() async {
    final stream = _remoteStream;
    _remoteStream = null;
    renderer.srcObject = null;

    if (stream != null) {
      for (final track in stream.getTracks()) {
        await track.stop();
      }
      await stream.dispose();
    }

    final pc = _peerConnection;
    _peerConnection = null;
    if (pc != null) {
      await pc.close();
      await pc.dispose();
    }
  }

  /// Closes the peer connection and renderer. Does not send STOP_STREAMING.
  Future<void> close() async {
    if (_disposed) return;
    _disposed = true;
    _connectGeneration++;
    await _teardownPeer();
    if (_rendererReady) {
      await renderer.dispose();
      _rendererReady = false;
    }
    dispose();
  }
}

File: vigia-web/src/app/core/services/streaming/whep-live-session.ts
export type WhepLiveStatus = 'connecting' | 'playing' | 'error';

export interface WhepLiveSessionState {
  status: WhepLiveStatus;
  errorMessage: string | null;
  isPaused: boolean;
  isClosed: boolean;
  remoteStream: MediaStream | null;
}

type StateListener = (state: WhepLiveSessionState) => void;

/**
 * Manages a receive-only WHEP session against MediaMTX.
 * Port of vigia_ui/lib/data/services/whep_live_session.dart.
 */
export class WhepLiveSession {
  private whepUrl: string;
  private status: WhepLiveStatus = 'connecting';
  private errorMessage: string | null = null;
  private isPaused = false;
  private isClosedFlag = false;
  private remoteStream: MediaStream | null = null;
  private peerConnection: RTCPeerConnection | null = null;
  private connectGeneration = 0;
  private listeners = new Set<StateListener>();

  constructor(whepUrl: string) {
    this.whepUrl = whepUrl;
  }

  getState(): WhepLiveSessionState {
    return {
      status: this.status,
      errorMessage: this.errorMessage,
      isPaused: this.isPaused,
      isClosed: this.isClosedFlag,
      remoteStream: this.remoteStream,
    };
  }

  subscribe(listener: StateListener): () => void {
    this.listeners.add(listener);
    listener(this.getState());
    return () => this.listeners.delete(listener);
  }

  beginConnecting(): void {
    if (this.isClosedFlag) {
      return;
    }
    this.status = 'connecting';
    this.errorMessage = null;
    this.isPaused = false;
    this.notify();
  }

  markError(error: unknown): void {
    if (this.isClosedFlag) {
      return;
    }
    this.status = 'error';
    this.errorMessage = error instanceof Error ? error.message : String(error);
    this.notify();
  }

  async connect(
    maxAttempts = 24,
    initialDelayMs = 500,
    maxDelayMs = 5000,
  ): Promise<void> {
    if (this.isClosedFlag) {
      return;
    }

    const generation = ++this.connectGeneration;
    this.beginConnecting();

    let delayMs = initialDelayMs;
    let lastError: unknown = null;

    for (let attempt = 0; attempt < maxAttempts; attempt++) {
      if (this.isClosedFlag || generation !== this.connectGeneration) {
        return;
      }

      try {
        await this.negotiate();
        if (this.isClosedFlag || generation !== this.connectGeneration) {
          return;
        }
        this.status = 'playing';
        this.errorMessage = null;
        this.notify();
        return;
      } catch (error) {
        lastError = error;
        await this.teardownPeer();
        if (this.isClosedFlag || generation !== this.connectGeneration) {
          return;
        }
        if (attempt === maxAttempts - 1) {
          break;
        }
        await this.delay(delayMs);
        delayMs = Math.min(delayMs * 2, maxDelayMs);
      }
    }

    if (this.isClosedFlag || generation !== this.connectGeneration) {
      return;
    }
    this.markError(lastError ?? 'Failed to connect to live stream');
  }

  togglePause(): void {
    const stream = this.remoteStream;
    if (!stream || this.status !== 'playing') {
      return;
    }

    this.isPaused = !this.isPaused;
    for (const track of stream.getVideoTracks()) {
      track.enabled = !this.isPaused;
    }
    for (const track of stream.getAudioTracks()) {
      track.enabled = !this.isPaused;
    }
    this.notify();
  }

  async close(): Promise<void> {
    if (this.isClosedFlag) {
      return;
    }
    this.isClosedFlag = true;
    this.connectGeneration++;

    this.status = 'connecting';
    this.errorMessage = null;
    this.isPaused = false;
    this.notify();

    await this.teardownPeer();
    await this.delay(0);
  }

  private async negotiate(): Promise<void> {
    const pc = new RTCPeerConnection({
      iceServers: [{ urls: 'stun:stun.l.google.com:19302' }],
    });
    this.peerConnection = pc;

    pc.ontrack = (event: RTCTrackEvent) => {
      if (this.isClosedFlag || !event.streams[0]) {
        return;
      }
      this.remoteStream = event.streams[0];
      this.notify();
    };

    pc.addTransceiver('video', { direction: 'recvonly' });
    pc.addTransceiver('audio', { direction: 'recvonly' });

    const offer = await pc.createOffer();
    await pc.setLocalDescription(offer);
    await this.waitForIceGathering(pc);

    const local = pc.localDescription;
    const sdp = local?.sdp;
    if (!sdp) {
      throw new Error('Empty local SDP offer');
    }

    const response = await fetch(this.whepUrl, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/sdp',
        Accept: 'application/sdp',
      },
      body: sdp,
    });

    if (response.status !== 200 && response.status !== 201) {
      throw new Error(`WHEP negotiate failed (${response.status})`);
    }

    const answerSdp = await response.text();
    if (!answerSdp) {
      throw new Error('Empty WHEP SDP answer');
    }

    await pc.setRemoteDescription({ type: 'answer', sdp: answerSdp });
  }

  private async waitForIceGathering(pc: RTCPeerConnection): Promise<void> {
    if (pc.iceGatheringState === 'complete') {
      return;
    }

    await new Promise<void>((resolve) => {
      const timeout = window.setTimeout(resolve, 8000);
      pc.onicegatheringstatechange = () => {
        if (pc.iceGatheringState === 'complete') {
          window.clearTimeout(timeout);
          resolve();
        }
      };
    });
  }

  private async teardownPeer(): Promise<void> {
    const stream = this.remoteStream;
    this.remoteStream = null;

    if (stream) {
      for (const track of stream.getTracks()) {
        track.stop();
      }
    }

    const pc = this.peerConnection;
    this.peerConnection = null;
    if (pc) {
      pc.close();
    }
  }

  private notify(): void {
    const state = this.getState();
    for (const listener of this.listeners) {
      listener(state);
    }
  }

  private delay(ms: number): Promise<void> {
    return new Promise((resolve) => window.setTimeout(resolve, ms));
  }
}
